src/ILLink.Tasks/CreateRuntimeRootDescriptorFile.cs

```diff
@@ -239,83 +299,130 @@ void OutputXml (string iLLinkTrimXmlFilePath, string outputFileName)
+				}
+
+				if (!members.keepAllFields && (members.fields != null)) {
+					foreach (string field in members.fields) {
+						XmlNode fieldNode = doc.CreateElement ("field");
+						XmlAttribute fieldName = doc.CreateAttribute ("name");
+						fieldName.Value = field;
+						fieldNode.Attributes.Append (fieldName);
+						typeNode.AppendChild (fieldNode);
 					}
+				}
 
-					if (members.methods != null) {
-						foreach (string method in members.methods) {
-							XmlNode methodNode = doc.CreateElement ("method");
-							XmlAttribute methodName = doc.CreateAttribute ("name");
-							methodName.Value = method;
-							methodNode.Attributes.Append (methodName);
-							typeNode.AppendChild (methodNode);
-						}
+				if (members.methods != null) {
+					foreach (string method in members.methods) {
+						XmlNode methodNode = doc.CreateElement ("method");
+						XmlAttribute methodName = doc.CreateAttribute ("name");
+						methodName.Value = method;
+						methodNode.Attributes.Append (methodName);
+						typeNode.AppendChild (methodNode);
 					}
 				}
-				assemblyNode.AppendChild (typeNode);
 			}
-			doc.Save (outputFileName);
+			assemblyNode.AppendChild (typeNode);
 		}
 
-		void AddClass (string classNamespace, string className, string classId, bool keepAllFields = false)
+
+		void AddClass (string classNamespace, string className, string classId, bool keepAllFields = false, FeatureSwitchMembers? featureSwitch = null)
 		{
 			string fullClassName = GetFullClassName (classNamespace, className);
 			if (fullClassName != null) {
 				if ((classId != null) && (classId != "NoClass")) {
 					classIdsToClassNames[classId] = fullClassName;
 				}
-				if (!classNamesToClassMembers.TryGetValue (fullClassName, out ClassMembers members)) {
-					members = new ClassMembers ();
-					classNamesToClassMembers[fullClassName] = members;
+
+				ClassMembers members;
+				if (!featureSwitch.HasValue) {
+					if (!classNamesToClassMembers.TryGetValue (fullClassName, out members)) {
+						members = new ClassMembers ();
+						classNamesToClassMembers[fullClassName] = members;
+					}
+				} else {
+					Dictionary<string, ClassMembers> currentFeatureSwitchMembers = featureSwitchMembers[featureSwitch.Value];
+					if (!currentFeatureSwitchMembers.TryGetValue (fullClassName, out members)) {
+						members = new ClassMembers ();
+						currentFeatureSwitchMembers[fullClassName] = members;
+					}
 				}
 				members.keepAllFields |= keepAllFields;
 			}
 		}
 
-		void AddField (string fieldName, string classId)
+		void AddField (string fieldName, string classId, FeatureSwitchMembers? featureSwitch = null)
 		{
 			string className = classIdsToClassNames[classId];
+			ClassMembers members;
 
-			ClassMembers members = classNamesToClassMembers[className];
+			if (!featureSwitch.HasValue) {
+				members = classNamesToClassMembers[className];
+			} else {
+				members = featureSwitchMembers[featureSwitch.Value][className];
+			}
```

Comment:
Would it be worth combining featureSwitchMembers and classNamesToClassMembers into a Dictionary<FeatureSwitchMembers?, ...> where a null key is just the default (no feature settings)? That would simplify some of the dictionary logic.

Reply:
If you don't have a strong objection, I prefer to keep these 2 separate for the time being. Should we re-visit feature switch to add member level support, I expect the design to be revamped and revisit this.